File: grammar/translator.py
import pgf
import sys

# comment out one of these:
#more precise, slower
#pgfm = "CSETranslator"

#less precise, faster
#pgfm = "CSEShallow"

#pgff = pgfm + ".pgf"
#ceng = pgfm + "Eng"
#cswe = pgfm + "Swe"

def trim(s0):
    s = s0.strip()
    if not(s):
        return s
    s = (s[0].lower() + s[1:])
    r = ""
    for c in s:
        if c in [',','.',':','(',')']:
            r = r + ' ' + c + ' '
        else:
            r = r + c
    return r

def main():
  if len(sys.argv) != 4:
    print("  usage: python3 translator.py <pgf-file-prefix> <from-lang-suffix> <to-lang-suffix>")
    print("  e.g. python3 translator.py CSETranslator Eng Swe")
    print("  The program reads and writes stdio line by line, e.g")
    print("    cat ../course_plans/TIN214Eng.txt | python3 translator.py CSEShallow Eng Swe")
    return
  pgfm = sys.argv[1]
  pgff = pgfm + ".pgf"
  ceng = pgfm + sys.argv[2]
  cswe = pgfm + sys.argv[3]
  gr = pgf.readPGF(pgff)
  eng = gr.languages[ceng]
  swe = gr.languages[cswe]
  print("#","translating with",pgff,"from",ceng,"to",cswe)
  while True:
    try:
        line = input("")
    except EOFError:
        break
    t = trim(line)
    if not(t):
        pass
    else:
          try:
              px = eng.parse(t)
              p,e = px.__next__()
              print(swe.linearize(e))
          except pgf.ParseError:
              print("# NO PARSE", t)

main()

File: grammar/Translator.hs
module Main where

import PGF2

import Data.Char
import Data.List
import qualified Data.Map as M
import System.Environment (getArgs)

-- AR May 2020

-- find translations of full sentences
--   cat ../course_plans/TIN214Eng.txt | runghc Translator.hs CSEShallow Eng Swe -debug
-- or just occurrences of terms:
--   cat ../course_plans/TIN214Swe.txt | runghc Translator.hs CSETranslator Swe Eng -cohorts

main = do
  xx <- getArgs
  case xx of
    mo:from:to:xx -> do
      env <- getEnv mo from to (elem "-debug" xx)
      let oper = case xx of
            _ | elem "-cohorts" xx -> cohorts 
            _ -> translate
      interact (unlines . map (oper env . trim) . lines)

    _ -> putStrLn usage

usage = "usage: runghc Translator.hs <pgf-file-prefix> <from-lang-suffix> <to-lang-suffix> -debug? -cohorts?"

trim s = case s of
  c:cs -> toLower c : tok cs
  _ -> s
 where
  tok s = case s of
    c:cs | elem c ".,:;()" -> ' ':c:' ':tok cs
    c:cs -> c : tok cs
    _ -> s

getEnv mo from to debug = do
  let pgff = mo ++ ".pgf"
  let ceng = mo ++ from
  let cswe = mo ++ to
  pgf <- readPGF pgff
  let Just eng = M.lookup ceng (languages pgf)
  let Just swe = M.lookup cswe (languages pgf)
  return (pgf,eng,swe,debug)

translate (pgf,eng,swe,debug) s = case parse eng (startCat pgf) s of
  ParseOk ((tree,_):_) -> linearize swe tree  ++ deb ("TREE " ++ show tree) 
  _ | all isSpace s -> s
  _ -> "# NO PARSE " ++ s ++ deb (miss s)
 where
   deb s = if debug then "\n# " ++ s else ""
   miss s = unwords $ "MISSING:" : [w | w <- words s, null (lookupMorpho eng w)]

cohorts (_,eng,swe,_) s = unlines $ ("# " ++ s): map prCohort funs
  where
   cohs = lookupCohorts eng s
   funs = nub [((beg,end),(c,f,linearize swe tree)) |
     (beg,c,ms,end) <- cohs,
     isInfixOf (words c) (words s),
     (f,_,_) <- ms, isSuffixOf "CSE" f,
     Just tree <- [readExpr f]
     ]
   prCohort ((beg,end),(c,f,lin)) = concat $ intersperse "\t" [show beg, show end, c, f, lin]
